### PR TITLE
Installation & Set Up Rucio ➟ Automation scripts for install and set up

### PR DESCRIPTION
**Pull request by airvzxf**
This is a proposal to improve the installation and configuration of Rucio, since at first, it experienced problems, especially when creating containers before the previous ones were finished. I decided to automate employing a Bash script on Linux for the installation and configuration of Rucio.

This proposal is justified because automation facilitates the creation of development environments quickly and efficiently. This will help the “[Namespace synchronization for shared Open Data](https://hepsoftwarefoundation.org/gsoc/2023/proposal_Rucio_OpenData.html)” project.

Here, I leave an animation in Asciinema exemplifying such automated tasks. In total, it takes about 9 minutes to finish.

[Rucio - k8s-tutorial - Automation scripts - From scratch](https://asciinema.org/a/566941)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,6 +62,9 @@ _NOTE: Before executing the following commands, please change directory to the c
 
 _NOTE: Replace the pod IDs with the ones from your instance, they change every time_
 
+_NOTE: You can execute the Linux script for better installation and understanding:
```
Comment (mlassnig) [inline README.md:65]: the -> this
better -> easier

File: README.md [hunk 2]
```diff
@@ -128,6 +131,9 @@ For Windows:
 
 ## Rucio usage
 
+_NOTE: You can execute the Linux script for better set up and understanding:
```
Comment (mlassnig) [inline README.md:134]: same as above

File: automation/01-install-rucio.bash [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+# Possible values for memory: max | format: <number>[<unit>], where unit = b, k, m or g, e.g., 4000mb
+MINIKUBE_MEMORY_RAM=8000mb
+# Possible values for CPUs: max | 1, 2, 3, 4, 5, ...
+MINIKUBE_CPU=4
+
+echo "# --------------------------------------"
+echo "# Check installed packages"
+echo "# --------------------------------------"
+KUBECTL_SUCCESS="The kubectl package is installed."
+KUBECTL_ERROR="The kubectl package is not installed. Please, follow this guide https://kubernetes.io/docs/tasks/tools/install-kubectl/"
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:14]: no comma after please :-p

File: automation/01-install-rucio.bash [hunk 4]
```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env bash
+set -e
+
+# Minikube
+# Possible values for memory: max | format: <number>[<unit>], where unit = b, k, m or g, e.g., 4000mb
+MINIKUBE_MEMORY_RAM=8000mb
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:6]: is this possible to override with env variables?

File: automation/01-install-rucio.bash (hunk 3 above)
Comment (mlassnig) [reply automation/01-install-rucio.bash:14]: and below

File: automation/01-install-rucio.bash [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+MINIKUBE_ERROR="The minikube package is not installed. Please, follow this guide https://minikube.sigs.k8s.io/docs/start/"
+type minikube &>/dev/null && echo "${MINIKUBE_SUCCESS}" || echo "${MINIKUBE_ERROR}"
+
+HELM_SUCCESS="The helm package is installed."
+HELM_ERROR="The helm package is not installed. Please, follow this guide https://helm.sh/docs/intro/install/"
+type helm &>/dev/null && echo "${HELM_SUCCESS}" || echo "${HELM_ERROR}"
+
+echo ""
+echo "# --------------------------------------"
+echo "# Celan local environment"
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:27]: Clean

File: automation/01-install-rucio.bash [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+minikube status | grep "Stopped" &>/dev/null && {
+  echo "${MINIKUBE_HAS_KUBERNETES}"
+  minikube status || true
+  read -rp "Do you want to delete the local Kubernetes clusters? (y/N): " WILL_DELETE_MINIKUBE
+}
+if [[ "${WILL_DELETE_MINIKUBE,,}" == "y" ]]; then
+  minikube delete --all=true
+fi
+
+MINIKUBE_HAS_NOT_PROFILE="The minikube not has profile. It will create a new one."
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:55]: The minikube does not have a profile.

File: automation/01-install-rucio.bash [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+minikube status | grep "Stopped" &>/dev/null && {
+  echo "${MINIKUBE_HAS_KUBERNETES}"
+  minikube status || true
+  minikube start --memory="${MINIKUBE_MEMORY_RAM}" --cpus="${MINIKUBE_CPU}"
+}
+
+echo "┌────────────────────────┐"
+echo "⟾ Helm: Install Postgres │"
+echo "└────────────────────────┘"
+KUBECTL_HAS_PVC="The kubectl has the Postgres PVC. It will delete."
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:119]: It will delete. -> Deleting.

File: automation/01-install-rucio.bash [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+  echo "${KUBECTL_HAS_PVC}"
+  kubectl delete pvc data-postgres-postgresql-0
+}
+helm delete postgres 2>/dev/null || true
+helm install postgres bitnami/postgresql -f ../postgres_values.yaml
+
+echo "┌──────────────────────────┐"
+echo "⟾ kubectl: Postgres set up │"
+echo "└──────────────────────────┘"
+echo "⤑ It will wait until the Postgres is set up. It could take several seconds or minutes."
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:130]: It will wait until ...   -> Waiting until

File: automation/01-install-rucio.bash [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+done
+echo ""
+kubectl get pods postgres-postgresql-0 || true
+
+echo "┌─────────────────────────────────┐"
+echo "⟾ kubectl: Rucio - Init Container │"
+echo "└─────────────────────────────────┘"
+kubectl delete pod init 2>/dev/null || true
+kubectl apply -f ../init-pod.yaml
+echo "⤑ It will wait until the Rucio Init Container is set up. It could take several seconds or minutes."
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:144]: same

File: automation/01-install-rucio.bash [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+echo "┌───────────────────────┐"
+echo "⟾ kubectl: Logs for FTS │"
+echo "└───────────────────────┘"
+kubectl logs "${FTS_SERVER_POD}"
+
+echo "┌───────────────────────┐"
+echo "⟾ helm: Install Daemons │"
+echo "└───────────────────────┘"
+echo "⤑ It will wait until the Daemons are set up. It could take several seconds or minutes."
```
Comment (mlassnig) [inline automation/01-install-rucio.bash:260]: same

**Review by mlassnig — COMMENTED** [review]
THanks @airvzxf, very nice. Just some minor typos in the comments below.

File: README.md (hunk 2 above)
Comment (airvzxf) [reply README.md]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.
Fixed in the commit 508e8d596246911768f13810b48d7085a1600ffb.

File: automation/01-install-rucio.bash (hunk 3 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 5 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 6 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 7 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 8 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 9 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 10 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit 84bf9a925933eebb4b73482beb79825180033d7d.

File: automation/01-install-rucio.bash (hunk 4 above)
Comment (airvzxf) [reply automation/01-install-rucio.bash]: Fixed in the commit fb245e3199a58021adda29142e66c639e8c64d06.

Furthermore, I added a help parameter for the script: `./01-install-rucio.bash -h` or `./01-install-rucio.bash --help`.

```text
▄▄▄▄▄▄▄▄▄▄▄▄
█   HELP   █
▀▀▀▀▀▀▀▀▀▀▀▀

Usage ▶ ./01-install-rucio.bash

Minikube
════════
Minikube accepts two parameters, the amount of memory and the number of CPUs.
  Memory:
    - max
    - format: <number>[<unit>], where unit = b, k, m or g, e.g., 4000mb
  CPUs:
    - max
    - 1, 2, 3, 4, 5, …, e.g., 8

  Usage
  ━━━━━
  To set up these parameters, need to use environment variables: MINIKUBE_MEMORY or MINIKUBE_CPU.
  Usage ▶ export MINIKUBE_MEMORY=5000mb MINIKUBE_CPU=3; ./01-install-rucio.bash
  Usage ▶ export MINIKUBE_CPU=max; ./01-install-rucio.bash
  Usage ▶ export MINIKUBE_MEMORY=max; ./01-install-rucio.bash

  Clean environment variables
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━
  The unset command, removes the environment variable
  Usage ▶ unset MINIKUBE_MEMORY MINIKUBE_CPU

  Actual value for environment variables
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
  Currently, these are the parameters which will set in the minikube. Blank means not arguments.
  ▶ minikube start --cpus=max
```

**Comment by airvzxf** [comment]
I fixed all the comments. Also, add a help menu by passing the `-h` or `--help` parameters in the script.

Any comment or improvement is welcome, I will gladly do it.

**Comment by rcarpa** [comment]
I find that the initiative is very good. I was just following the tutorial myself, before even seeing this PR and thought "this will not work if you don't wait enough between commands". 

Nevertheless, I would vote against this change. Using bash scripts to orchestrate kubernetes shouldn't be encouraged. Not even in a tutorial. There are much better existing, of-the-shelf, tools for that. Terraform is the first which comes to mind. Or flux. Or argocd. I'm sure there are other tools, which may be even better suited for a tutorial.

**Comment by airvzxf** [comment]
Hi, I am open to any possibility. That's true, it should be better with some orchestrate tool for Kubernetes. Since this task is for the GSoC, I was in a difficult decision between spending a lot of time on something “perfect/correct” and creating something to demonstrate some improvement then resolve it. I decided to go for the second option.

I have two ideas for this: the first is revoking the merge request and creating a To-Do list with this task: “add an orchestrate tool for Kubernetes”. The second option is merging this and adding a note about the improvement, replacing the Bash script for some orchestrate tool for Kubernetes. In both cases, it opens the opportunity for someone to apply this orchestrates.

**Comment by rcarpa** [comment]
Lets merge it. I'll search for some better solution when I'll have time later

**Pull request merged**